dags/synaptor_dags.py

```diff
@@ -43,324 +46,208 @@
 
-generate_merge_ccs = generate_op(fileseg_dag, "merge_ccs", image=SYNAPTOR_IMAGE)
-wait_merge_ccs = wait_op(fileseg_dag, "merge_ccs")
+def fill_dag(dag: DAG, tasklist: list[Task], collect_metrics: bool = True) -> DAG:
+    """Fills a synaptor DAG from a list of Tasks."""
+    drain = drain_op(dag)
+    init_cloudvols = manager_op(dag, "init_cloudvols", image=SYNAPTOR_IMAGE)
 
-generate_remap = generate_op(fileseg_dag, "remap", image=SYNAPTOR_IMAGE)
-wait_remap = wait_op(fileseg_dag, "remap")
+    drain >> init_cloudvols
 
+    if collect_metrics:
+        metrics = collect_metrics_op(dag)
+        metrics >> drain
 
-# DEPENDENCIES
-# Drain old tasks before doing anything
-collect_metrics_op(fileseg_dag) >> drain >> sanity_check >> init_cloudvols  # >> generate_ngl_link
+    curr_cluster = ""
+    curr_operator = init_cloudvols
+    for task in tasklist:
+        if isinstance(task, CPUTask) and "cpu" not in curr_cluster:
+            curr_cluster, curr_operator = change_clusters(
+                dag, curr_cluster, "synaptor-cpu", curr_operator
+            )
+        elif isinstance(task, GPUTask) and "gpu" not in curr_cluster:
+            curr_cluster, curr_operator = change_clusters(
+                dag, curr_cluster, "synaptor-gpu", curr_operator
+            )
+        elif isinstance(task, GraphTask) and "graph" not in curr_cluster:
+            curr_cluster, curr_operator = change_clusters(
+                dag, curr_cluster, "synaptor-seggraph", curr_operator
+            )
 
-# Worker dag
-(init_cloudvols >> scale_up_cluster >> workers >> scale_down_cluster)
+        curr_operator = add_task(dag, task, curr_operator)
 
-# Generator/task dag
-(
-    init_cloudvols
-    # >> sanity_check_report
-    >> generate_chunk_ccs
-    >> wait_chunk_ccs
-    >> generate_merge_ccs
-    >> wait_merge_ccs
-    >> generate_remap
-    >> wait_remap
-    >> generate_self_destruct
-    >> wait_self_destruct
-)
+    assert curr_cluster is not None, "no tasks?"
 
+    return dag
 
-# =========================================
-# DB segmentation
-# "run synaptor db segmentation"
-#
-# This performs the "merge_ccs" task above in separate extra steps (in parallel)
-#
-# I will also soon add some database management commands (creating indexes)
-
-dbseg_dag = DAG(
-    "synaptor_db_seg",
-    default_args=default_args,
-    schedule_interval=None,
-    tags=["synaptor"],
-)
 
+def change_clusters(
+    dag: DAG, prev_cluster_tag: str, next_cluster_key: str, curr_operator: Operator,
+) -> tuple[str, Operator]:
+    # scaling down previous cluster
+    scale_down, curr_operator = scale_down_cluster(dag, prev_cluster_tag, curr_operator)
 
-# OP INSTANTIATION
-# Cluster management
-drain = drain_op(dbseg_dag)
-sanity_check = manager_op(dbseg_dag, "sanity_check", image=SYNAPTOR_IMAGE)
-init_cloudvols = manager_op(dbseg_dag, "init_cloudvols", image=SYNAPTOR_IMAGE)
-init_db = manager_op(dbseg_dag, "init_db", image=SYNAPTOR_IMAGE)
+    # generating a new cluster tag
```

Comment:
I am confused, I thought you know exactly the number of waves needed for your workflow. If it is important for the label to be meaningful, do you have better options, I mean something like "scale_operator_assignment" instead of "scale_operator_1".
By "suspicious situations" I am referring to your check below `if guess_i == 5`. When would this happen?

Reply:
I do know the number of waves needed for every workflow I've added so far. I wanted this function to support other lists of tasks that someone else might write (e.g., nucleus segmentation and overlap) without someone needing to think through the number of times you need to spin up each cluster. It's not likely for someone to write a list of tasks that triggers this, but spinning any cluster up 5 times seems like a clear sign that something was wrong.

Reply:
I'll just take it out. Not worth it.

Reply:
OK, I suppose in the longer term, it make sense to factor out the cluster related operators and put them in a common place, it will make it harder to create meaningful and unique task ids.